pkg/workspace/workspace.go

```diff
@@ -195,8 +195,7 @@ func (ws *Workspace) ListCatalogs() ([]*string, wfapi.Error) {
 }
 
 // Get a catalog ware from a workspace, doing lookup by CatalogRef.
-// This will first check all catalogs within the "catalogs" subdirectory, if it exists
-// then, it will check the "catalog" subdirectory, if it exists
+// This will first search catalogs within the "catalogs" subdirectory and then search the "catalog" subdirectory
```

Comment:
Hm.  Okay, feel free to deflect this as "not in this PR", but...

Now that I'm looking at it, this function below is doing kind of a silly thing.

The behavior I think would be most excellent would be to check the nearest catalog first, then check any other catalogs seen on the way to the root and make sure they match, and then ultimately do the range over root catalogs and make sure it exists in exactly one of them and still matches.

(Or more realistically, that should probably be a few different functions, or, just one function with a bunch of flags to control how much work it invests in active sanity checks.)

Reply:
>check the nearest catalog first, then check any other catalogs seen on the way to the root and make sure they match

This is something that could work for the `WorkspaceSet` implementation. For a single workspace it might not make sense. However you could have it only check `catalogs` directory if it's a root workspace if that's a dichotomy we want to enforce.